server/src/models/mod.rs

```diff
@@ -75,6 +84,87 @@ impl PartialOrd<ErrorBound> for f32 {
     }
 }
 
+/// Model that uses the fewest number of bytes per value.
+pub struct SelectedModel {
+    /// Id of the model type that created this model.
+    pub model_type_id: u8,
+    /// The selected model's length.
+    pub end_index: usize,
+    /// The selected model's minimum value.
+    pub min_value: Value,
+    /// The selected model's maximum value.
+    pub max_value: Value,
+    /// Data required in addition to `self.min` and `self.max` for the model to
+    /// reconstruct the values it represents when given a specific timestamp.
+    pub values: Vec<u8>,
+}
+
+impl SelectedModel {
+    fn new(
+        model_type_id: u8,
+        end_index: usize,
+        min_value: Value,
+        max_value: Value,
+        values: Vec<u8>,
+    ) -> Self {
+        Self {
+            model_type_id,
+            end_index,
+            min_value,
+            max_value,
+            values,
+        }
+    }
+}
+
+/// Select the model that uses the fewest number of bytes per value.
+pub fn select_model(
+    start_index: usize,
+    pmc_mean: PMCMean,
+    swing: Swing,
+    gorilla: Gorilla,
+    uncompressed_values: &ValueArray,
+) -> SelectedModel {
+    let bytes_per_value = [
+        (PMC_MEAN_ID, pmc_mean.get_bytes_per_value()),
+        (SWING_ID, swing.get_bytes_per_value()),
+        (GORILLA_ID, gorilla.get_bytes_per_value()),
+    ];
+
+    // unwrap() cannot fail as the array is not empty and there are no NaN.
+    let model_type_id = bytes_per_value
+        .iter()
+        .min_by(|x, y| f32::partial_cmp(&x.1, &y.1).unwrap())
+        .unwrap()
+        .0;
+
+    match model_type_id {
+        PMC_MEAN_ID => {
+            let value = pmc_mean.get_model();
+            let end_index = start_index + pmc_mean.get_length() - 1;
+            SelectedModel::new(PMC_MEAN_ID, end_index, value, value, vec![])
+        }
+        SWING_ID => {
+            let (min_value, max_value) = swing.get_model();
+            let end_index = start_index + swing.get_length() - 1;
+            SelectedModel::new(SWING_ID, end_index, min_value, max_value, vec![])
+        }
+        GORILLA_ID => {
+            let end_index = start_index + gorilla.get_length() - 1;
+            let uncompressed_values = &uncompressed_values.values()[start_index..end_index];
+            let min_value = uncompressed_values
+                .iter()
+                .fold(Value::MAX, |x, y| Value::min(x, *y));
```

Comment:
The opposite value is used as the initial value to ensure it is always replaced, e.g, if `Value::MIN` was used for `min_value`, the result would always be `Value::MIN` as a smaller value cannot exist. However, as we also consider `Value::NAN` an actual value both should be initialized with `Value::NAN` so it is returned instead of `Value::MIN` or `f32::Value` if all of the values in the segment is `Value::NAN`. Thus, both are now initialized with `Value::NAN` and tests have been added that checks that the correct minimum and maximum values are returned.